cpp/src/arrow/array.h

```diff
@@ -628,6 +659,8 @@ class ARROW_EXPORT UnionArray : public Array {
   /// Only use this while the UnionArray is in scope
   const Array* UnsafeChild(int pos) const;
 
+  std::shared_ptr<DataType> value_type(int pos) const;
```

Comment:
It wasn't initially clear that this is the child type, maybe call this `child_type` instead (and add doxygen comment)?

Reply:
This is now gone and using the Array's type to get this information. It's more consistent with how things are done for ListArray and StructArray.